ReactWindows/ReactNative.Shared/Bridge/ReactContextNativeModuleBase.cs

```diff
@@ -36,32 +36,37 @@ protected ReactContextNativeModuleBase(ReactContext reactContext, IActionQueue a
         }
 
         /// <summary>
-        /// Instantiates the <see cref="ReactContextNativeModuleBase"/>.
+        /// The action queue used by the native module.
         /// </summary>
-        /// <param name="reactContext">The React context.</param>
-        /// <param name="delegateFactory">The React method delegate factory.</param>
-        protected ReactContextNativeModuleBase(ReactContext reactContext, IReactDelegateFactory delegateFactory)
-            : base(delegateFactory)
+        /// <remarks>
+        /// Defaults to the native modules action queue. If you override this getter,
+        /// be sure to return the same IActionQueue instance each time this getter is
+        /// accessed for a given module instance.
+        /// </remarks>
+        public virtual IActionQueue ActionQueue
         {
-            if (reactContext == null)
-                throw new ArgumentNullException(nameof(reactContext));
-
-            Context = reactContext;
+            get
+            {
+                return Context.IndisposableNativeModulesQueue;
+            }
         }
 
         /// <summary>
-        /// Instantiates the <see cref="ReactContextNativeModuleBase"/>.
+        /// Disposes the ActionQueue.
         /// </summary>
-        /// <param name="reactContext">The React context.</param>
-        /// <param name="delegateFactory">The React method delegate factory.</param>
-        /// <param name="actionQueue">The action queue.</param>
-        protected ReactContextNativeModuleBase(ReactContext reactContext, IReactDelegateFactory delegateFactory, IActionQueue actionQueue)
-            : base(delegateFactory, actionQueue)
+        /// <remarks>
+        /// If you override <see cref="ActionQueue" /> to return a custom action
+        /// queue, you should also override this method to ensure your action
+        /// queue is disposed properly. This method runs on the native modules
+        /// queue.
+        /// 
+        /// Note that <see cref="INativeModule.OnReactInstanceDispose" />
+        /// isn't the appropriate place to dispose your action queue because
+        /// that means you'll dispose your action queue while running on
+        /// your action queue.
+        /// </remarks>
+        public virtual void DisposeActionQueue()
```

Comment:
Ugh - we're going to need a 40 page manual for writing native modules. I'm not a fan of the extra lifecycle hooks. I actually don't really like `OnReactInstanceDispose` either, ultimately I wanted to change everything to `IDisposable` or `IAsyncDisposable`

Reply:
Do you have an alternative suggestion?

I don't think we can dispose the queue for the user because shared queues shouldn't get disposed when a module gets disposed.

The challenge with reusing OnReactDispose for disposing the queue is then the queue will be disposed while running on the queue.

Reply:
If we take the approach of only using the interface to add an action queue (rather than forcing it onto the ReactContextNativeModuleBase), the user can decide of when they want to dispose. We can change the semantics of life cycle events to always use the core native module queue thread if that would help.

Reply:
Do you have an alternative suggestion?

There were several alternatives I considered and decided against:
  - RNW disposes the queue for the user when it disposes the module. The challenge is shared queues shouldn't get disposed when a module gets disposed.
  - The user is responsible for disposing the queue in `OnReactInstanceDispose`. The challenge is `OnReactInstanceDispose` runs on the queue so the queue's dispose method would run on the queue itself.

---
In reply to: [174975232](https://github.com/Microsoft/react-native-windows/pull/1650#discussion_r174975232) [](ancestors = 174975232)

Reply:
You can add the `IHasActionQueue` interface to whatever you want. Only `ReactContextNativeModuleBase` implements it by default.

How does not building `IHasActionQueue` into `ReactContextNativeModuleBase` give the user more flexibility around disposal?

I think moving the lifecycle events to the native modules queue would remove a value proposition of the `ActionQueue`. If your module has a custom `LimitedConcurrencyActionQueue`, then it'll have to worry about thread-safety in the module's lifecycle events. Either you'd have to use locks or dispatch to the `ActionQueue`. `OnReactInstanceDispose` sounds like it's possibly expected to complete all of its work synchronously in which case dispatching to the `ActionQueue` in your `OnReactInstanceDispose` implementation wouldn't be an option since dispatching is async.

---
In reply to: [175147495](https://github.com/Microsoft/react-native-windows/pull/1650#discussion_r175147495) [](ancestors = 175147495)

Reply:
We're overcomplicating ReactContextNativeModuleBase, which is used in many places and often doesn't require a custom action queue. Maybe it would be better to create another subclass that derives from ReactContextNativeModuleBase and implements IHasActionQueue.

Reply:
I can create an `ActionQueueNativeModuleBase` which inherits from `ReactContextNativeModuleBase` and exposes 
 `ActionQueue` and `DisposeActionQueue`. Is that what you want? I don't have a strong opinion about this.

I just want to be able to say `ActionQueue.Dispatch(...)` in my module and have it work regardless of whether or not the module is using the native module queue (you can't do this today because `ActionQueue` is null by default).